File: ios/Classes/Data/CustomErrorMessages.swift
// Copyright © 2022 Olo Inc. All rights reserved.
// This software is made available under the Olo Pay SDK License (See LICENSE.md file)
//
//  CustomErrorMessages.swift
//  olo_pay_sdk
//
//  Created by Richard Dowdy on 3/5/24.
//

import Foundation
import OloPaySDK

class CustomErrorMessages {
    private let invalidCardNumber: String?
    private let emptyCardNumber: String?
    private let invalidExpiration: String?
    private let emptyExpiration: String?
    private let invalidCvv: String?
    private let emptyCvv: String?
    private let invalidPostalCode: String?
    private let emptyPostalCode: String?
    let unsupportedCardNumber: String?
    
    init(customErrorMessages: NSDictionary?) {
        self.invalidCardNumber = (customErrorMessages?[OPCardField.number.flutterBridgeValue()] as? [String:Any])?[DataKeys.invalidErrorKey] as? String
        self.emptyCardNumber = (customErrorMessages?[OPCardField.number.flutterBridgeValue()] as? [String:Any])?[DataKeys.emptyErrorKey] as? String
        self.invalidExpiration = (customErrorMessages?[OPCardField.expiration.flutterBridgeValue()] as? [String:Any])?[DataKeys.invalidErrorKey] as? String
        self.emptyExpiration = (customErrorMessages?[OPCardField.expiration.flutterBridgeValue()] as? [String:Any])?[DataKeys.emptyErrorKey] as? String
        self.invalidCvv = (customErrorMessages?[OPCardField.cvv.flutterBridgeValue()] as? [String:Any])?[DataKeys.invalidErrorKey] as? String
        self.emptyCvv = (customErrorMessages?[OPCardField.cvv.flutterBridgeValue()] as? [String:Any])?[DataKeys.emptyErrorKey] as? String
        self.invalidPostalCode = (customErrorMessages?[OPCardField.postalCode.flutterBridgeValue()] as? [String:Any])?[DataKeys.invalidErrorKey] as? String
        self.emptyPostalCode = (customErrorMessages?[OPCardField.postalCode.flutterBridgeValue()] as? [String:Any])?[DataKeys.emptyErrorKey] as? String
        self.unsupportedCardNumber = customErrorMessages?[DataKeys.unsupportedCardErrorKey] as? String
    }
    
    init(cvvCustomErrors: NSDictionary?) {
        invalidCardNumber = nil
        emptyCardNumber = nil
        invalidExpiration = nil
        emptyExpiration = nil
        invalidCvv = cvvCustomErrors?[DataKeys.invalidErrorKey] as? String
        emptyCvv = cvvCustomErrors?[DataKeys.emptyErrorKey] as? String
        invalidPostalCode = nil
        emptyPostalCode = nil
        unsupportedCardNumber = nil
    }
    
    static func getDefaultErrorMessage(
        for cardField: OPCardField,
        with state: OPCardFieldStateProtocol,
        _ ignoreUneditedFields: Bool,
        _ cardBrand: OPCardBrand? = nil
    ) -> String {
        return getDefaultErrorMessage(ignoreUneditedFields, [cardField: state], cardBrand ?? OPCardBrand.unknown)
    }
    
    static func getDefaultErrorMessage(
        _ ignoreUneditedFields: Bool,
        _ cardFields: [OPCardField: OPCardFieldStateProtocol],
        _ cardBrand: OPCardBrand
    ) -> String {
        let errorFields = getErrorFields(ignoreUneditedFields, cardFields)

        if let numberState = errorFields[.number] {
            if numberState.isEmpty {
                return OPStrings.emptyCardNumberError
            } else if cardBrand == .unsupported {
                return OPStrings.unsupportedCardError
            } else {
                return OPStrings.invalidCardNumberError
            }
        }
        
        if let expirationState = errorFields[.expiration] {
            return expirationState.isEmpty ? OPStrings.emptyExpirationError : OPStrings.invalidExpirationError
        }
        
        if let cvvState = errorFields[.cvv] {
            return cvvState.isEmpty ? OPStrings.emptyCvvError : OPStrings.invalidCvvError
        }
        
        if let postalCodeState = errorFields[.postalCode] {
            return postalCodeState.isEmpty ? OPStrings.emptyPostalCodeError : OPStrings.invalidPostalCodeError
        }
            
        return ""
    }
    
    func getCustomErrorMessage(
        for cardField: OPCardField,
        with state: OPCardFieldStateProtocol,
        _ ignoreUneditedFields: Bool,
        _ cardBrand: OPCardBrand? = nil
    ) -> String? {
        return getCustomErrorMessage(ignoreUneditedFields, [cardField: state], cardBrand ?? OPCardBrand.unknown)
    }
    
    func getCustomErrorMessage(
        _ ignoreUneditedFields: Bool,
        _ cardFields: [OPCardField: OPCardFieldStateProtocol],
        _ cardBrand: OPCardBrand
    ) -> String? {
        let errorFields = CustomErrorMessages.getErrorFields(ignoreUneditedFields, cardFields)

        if let numberState = errorFields[.number] {
            if numberState.isEmpty {
                return emptyCardNumber
            } else if cardBrand == .unsupported {
                return unsupportedCardNumber
            } else {
                return invalidCardNumber
            }
        }
        
        if let expirationState = errorFields[.expiration] {
            return (expirationState.isEmpty ? emptyExpiration : invalidExpiration)
        }
        
        if let cvvState = errorFields[.cvv] {
            return (cvvState.isEmpty ? emptyCvv : invalidCvv)
        }
        
        if let postalCodeState = errorFields[.postalCode] {
            return (postalCodeState.isEmpty ? emptyPostalCode : invalidPostalCode)
        }
            
        return nil
    }
    
    static func getErrorFields(_ ignoreUneditedFields: Bool, _ cardState: [OPCardField: OPCardFieldStateProtocol]) -> [OPCardField: OPCardFieldStateProtocol] {
        if(!ignoreUneditedFields) {
            return cardState.filter({!$0.value.isValid})
        }
        
        return cardState.filter({!$0.value.isValid && $0.value.wasEdited && $0.value.wasFirstResponder})
    }
    
}

File: android/src/main/kotlin/com/olo/flutter/olo_pay_sdk/data/CustomErrorMessages.kt
// Copyright © 2022 Olo Inc. All rights reserved.
// This software is made available under the Olo Pay SDK License (See LICENSE.md file)
package com.olo.flutter.olo_pay_sdk.data

import com.olo.olopay.data.CardBrand
import com.olo.olopay.data.CardField
import com.olo.olopay.data.ICardFieldState

data class CustomErrorMessages constructor(
    val invalidCardNumber: String?,
    val emptyCardNumber: String?,
    val unsupportedCardNumber: String?,
    val invalidExpiration: String?,
    val emptyExpiration: String?,
    val invalidCvv: String?,
    val emptyCvv: String?,
    val invalidPostalCode: String?,
    val emptyPostalCode: String?
){
    constructor(customErrorMessagesMap: Map<*, *>): this(
        (customErrorMessagesMap[CardField.CardNumber.toString()] as? Map<*, *>)?.get(DataKeys.InvalidErrorKey) as? String,
        (customErrorMessagesMap[CardField.CardNumber.toString()] as? Map<*, *>)?.get(DataKeys.EmptyErrorKey) as? String,
        customErrorMessagesMap[DataKeys.UnsupportedCardErrorKey] as? String,
        (customErrorMessagesMap[CardField.Expiration.toString()] as? Map<*, *>)?.get(DataKeys.InvalidErrorKey) as? String,
        (customErrorMessagesMap[CardField.Expiration.toString()] as? Map<*, *>)?.get(DataKeys.EmptyErrorKey) as? String,
        (customErrorMessagesMap[CardField.Cvv.toString()] as? Map<*, *>)?.get(DataKeys.InvalidErrorKey) as? String,
        (customErrorMessagesMap[CardField.Cvv.toString()] as? Map<*, *>)?.get(DataKeys.EmptyErrorKey) as? String,
        (customErrorMessagesMap[CardField.PostalCode.toString()] as? Map<*, *>)?.get(DataKeys.InvalidErrorKey) as? String,
        (customErrorMessagesMap[CardField.PostalCode.toString()] as? Map<*, *>)?.get(DataKeys.EmptyErrorKey) as? String
    )

    constructor(invalidCvvError: String?, emptyCvvError: String?): this(
        null,
        null,
        null,
        null,
        null,
        invalidCvvError,
        emptyCvvError,
        null,
        null
    )

    fun getCustomErrorMessage(ignoreUneditedFieldErrors: Boolean, cardFields: Map<CardField, ICardFieldState>, cardBrand: CardBrand? = null): String? {
        val errorFields = getErrorFields(ignoreUneditedFieldErrors, cardFields)
        return when  {
            errorFields.containsKey(CardField.CardNumber) -> {
                val state = errorFields[CardField.CardNumber]!!
                if(state.isEmpty) {
                    emptyCardNumber
                } else if (cardBrand == CardBrand.Unsupported) {
                    unsupportedCardNumber
                } else {
                    invalidCardNumber
                }
            }
            errorFields.containsKey(CardField.Expiration) -> {
                if(errorFields[CardField.Expiration]!!.isEmpty) emptyExpiration else invalidExpiration
            }
            errorFields.containsKey(CardField.Cvv) -> {
                if (errorFields[CardField.Cvv]!!.isEmpty) emptyCvv else invalidCvv
            }
            errorFields.containsKey(CardField.PostalCode) -> {
                if (errorFields[CardField.PostalCode]!!.isEmpty) emptyPostalCode else invalidPostalCode
            }
            else -> {
                null
            }
        }
    }

    fun getErrorFields(ignoreUneditedFieldErrors: Boolean, cardFields: Map<CardField, ICardFieldState>): Map<CardField, ICardFieldState> {
        // Return all error fields regardless of state
        if (!ignoreUneditedFieldErrors)
            return cardFields.filterKeys {
                !cardFields[it]!!.isValid
            }

        return cardFields.filterKeys {
            val state = cardFields[it]!!
            !state.isValid && state.wasEdited && state.wasFocused
        }
    }
}
